copybookreader-plugins/src/main/java/co/cask/hydrator/plugin/batch/CopyBookRecordReader.java

```diff
@@ -0,0 +1,150 @@
+  private ExternalRecord externalRecord;
+  private int recordByteLength;
+  private long start;
+  private long position;
+  private long end;
+  private LongWritable key = null;
+  private Map<String, String> value = null;
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
+    // Get configuration
+    Configuration conf = context.getConfiguration();
+    int fileStructure = Integer.parseInt(conf.get(COPYBOOK_INPUTFORMAT_FILE_STRUCTURE,
+                                                  Integer.toString(DEFAULT_FILE_STRUCTURE)));
+    Path path = new Path(conf.get(COPYBOOK_INPUTFORMAT_DATA_HDFS_PATH));
+    FileSystem fs = FileSystem.get(path.toUri(), conf);
+    //create input stream for the cobol Copybook contents
+    InputStream inputStream = IOUtils.toInputStream(conf.get(COPYBOOK_INPUTFORMAT_CBL_CONTENTS), "UTF-8");
+    BufferedInputStream bufferedInputStream = new BufferedInputStream(inputStream);
+    try {
+      externalRecord = CopybookIOUtils.getExternalRecord(bufferedInputStream);
+      recordByteLength = CopybookIOUtils.getRecordLength(externalRecord, fileStructure);
+      org.apache.hadoop.mapreduce.lib.input.FileSplit fileSplit =
+        (org.apache.hadoop.mapreduce.lib.input.FileSplit) split;
+
+      start = fileSplit.getStart();
+      end = start + fileSplit.getLength();
+
+      BufferedInputStream fileIn = new BufferedInputStream(fs.open(fileSplit.getPath()));
+      /* Jump to the point in the split at which the first
+       whole record of split starts if not the first InputSplit*/
+      if (start != 0) {
+        position = start - (start % recordByteLength) + recordByteLength;
+        fileIn.skip(position);
+      }
+      reader = CopybookIOUtils.getAndOpenLineReader(fileIn, fileStructure, externalRecord);
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException {
+    AbstractLine line = reader.read();
+    if (line == null) {
+      return false;
+    }
+    if (position > end) {
+      return false;
+    }
+    if (key == null) {
+      key = new LongWritable();
+    }
+    value = new HashMap();
+    position += recordByteLength;
+    key.set(position);
+    for (ExternalField field : externalRecord.getRecordFields()) {
+      /*check if field is binary
+       in case of binary fields encode using Base64 format to preseve data consistency*/
+      // TODO : check int value for data types returned by jrecord to match java data types
+      // and create schema accrodindly
+      AbstractFieldValue filedValue = line.getFieldValue(field.getName());
+      if (filedValue.isBinary()) {
+        value.put(field.getName(), new String(Base64.decodeBase64(Base64.encodeBase64String(
+          filedValue.toString().getBytes()))));
+      } else {
+        value.put(field.getName(), filedValue.toString());
```

Comment:
We are returning Map<String, AbstractFieldValue> , since user can also define the required output schema and it will be available in config. Hence creating StructuredRecord in transform method.